test/e2e/experimental_features.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/utils/pointer"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	addonsv1 "sigs.k8s.io/cluster-api/exp/addons/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/cluster-api/test/framework/clusterctl"
+	"sigs.k8s.io/cluster-api/util"
+)
+
+// ExperimentalFeaturesSpecInput is the input for ExperimentalFeaturesSpec.
+type ExperimentalFeaturesSpecInput struct {
+	E2EConfig             *clusterctl.E2EConfig
+	ClusterctlConfigPath  string
+	BootstrapClusterProxy framework.ClusterProxy
+	ArtifactFolder        string
+	SkipCleanup           bool
+}
+
+// ExperimentalFeaturesSpec implements a spec that tests experimental featured that are gated with feature gates.
+// Once an experimental feature graduates, its test can be a separate spec.
+func ExperimentalFeaturesSpec(ctx context.Context, inputGetter func() ExperimentalFeaturesSpecInput) {
```

Comment:
If we go for using ClusterResourceSet for installing resources in the e2e tests, the additional test IMO is not necessary anymore.

Reply:
I'd prefer to keep e2e test for now and remove it if we use CRS for installing e2e resources. Also, there are several edge cases we can cover here, I'd have a seperate e2e test for each experiment until they are promoted.

Reply:
Created a follow up PR: https://github.com/kubernetes-sigs/cluster-api/issues/3292